src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts

```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
+				if (!ops.length) {
+					return []; // Nothing to do
+				}
+
+				return ops.map(op => new ResourceTextEdit(model.uri, { ...op, text: op.text || '' }, model.getVersionId()));
+			}));
+
+			const filteredEdits = allCellEdits.flat().filter(edit => edit !== undefined) as ResourceEdit[];
+			await this.bulkEditService.apply(filteredEdits, { label: localize('trimNotebookWhitespace', "Notebook Trim Trailing Whitespace"), code: 'undoredo.notebookTrimTrailingWhitespace' });
+
+		} finally {
+			progress.report({ increment: 100 });
+			disposable.dispose();
+		}
+	}
+}
+
+class TrimFinalNewLinesParticipant implements IStoredFileWorkingCopySaveParticipant {
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IEditorService private readonly editorService: IEditorService,
+		@ITextModelService private readonly textModelService: ITextModelService,
+		@IBulkEditService private readonly bulkEditService: IBulkEditService,
+	) { }
+
+	async participate(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>, _token: CancellationToken): Promise<void> {
+		if (this.configurationService.getValue<boolean>('files.trimTrailingWhitespace')) {
+			this.doTrimFinalNewLines(workingCopy, context, progress);
+		}
+	}
+
+	/**
+	 * returns 0 if the entire file is empty
+	 */
+	private findLastNonEmptyLine(model: ITextModel): number {
+		for (let lineNumber = model.getLineCount(); lineNumber >= 1; lineNumber--) {
+			const lineContent = model.getLineContent(lineNumber);
+			if (lineContent.length > 0) {
+				// this line has content
+				return lineNumber;
+			}
+		}
+		// no line has content
+		return 0;
+	}
+
+	private async doTrimFinalNewLines(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>): Promise<void> {
+		if (!workingCopy.model || !(workingCopy.model instanceof NotebookFileWorkingCopyModel)) {
+			return;
+		}
+
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+
+		let cannotTouchLineNumber = 0;
+		let viewCell: ICellViewModel | undefined = undefined;
+
+		// autosave -- don't trim entire line, only up to cursor, so need to track position of cursor(s)
+		if (context.reason === SaveReason.AUTO) {
+			viewCell = getNotebookViewCell(this.editorService);
+			if (!viewCell) {
+				return;
+			}
+			const selections = viewCell.getSelections();
+			for (const sel of selections) {
+				if (viewCell.model.textModel) {
+					cannotTouchLineNumber = Math.max(cannotTouchLineNumber, sel.selectionStartLineNumber);
+				}
+			}
+		}
+
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
+					return;
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
```

Comment:
We don't need `model` here, `cell.textBuffer` is sufficient.

Reply:
Model need not be specified within the file available .
-> only cell.textbuffer would be sufficient for the function in order to run